apps/src/lib/node/ledger/shell/finalize_block.rs

```diff
@@ -518,26 +515,33 @@ where
                         msg
                     );
 
-                    // If transaction type is Decrypted and failed because of
-                    // out of gas or invalid section commtiment, remove its hash
-                    // from storage to allow rewrapping it
+                    // If transaction type is Decrypted and didn't failed
+                    // because of out of gas nor invalid
+                    // section commitment, commit its hash to prevent replays
                     if let Some(wrapper) = embedding_wrapper {
-                        if matches!(
+                        if !matches!(
                             msg,
                             Error::TxApply(protocol::Error::GasError(_))
                                 | Error::TxApply(
                                     protocol::Error::MissingSection(_)
                                 )
+                                | Error::TxApply(
+                                    protocol::Error::ReplayAttempt(_)
+                                )
                         ) {
-                            self.allow_tx_replay(wrapper);
+                            self.commit_inner_tx_hash(wrapper);
+                        } else if let Error::TxApply(
+                            protocol::Error::ReplayAttempt(_),
+                        ) = msg
+                        {
+                            // Remove the wrapper hash but keep the inner tx
```

Comment:
Can we further explicate this comment to explain why this is safe? (as I understand it, because the inner transaction hash will still be in storage, and block any future replay)